pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -159,4 +181,271 @@ def is_root_lvl(patch_level):
+               "flux_z": "Vz",
+               "bulkVelocity_x": "Vx",
+               "bulkVelocity_y": "Vy",
+               "bulkVelocity_z": "Vz",
+               "density": "rho"}
+
+
+particle_files_patterns = ("domain", "patchGhost", "levelGhost")
+
+
+def is_particle_file(filename):
+    return any([pattern in filename for pattern in particle_files_patterns])
+
+
+
+def particle_dataset_name(basename):
+    """
+    return "alpha_domain" from ions_pop_ions_alpha_domain.h5
+    """
+    popname = basename.strip(".h5").split("_")[-2]
+    part_type = basename.strip(".h5").split("_")[-1]
+    dataset_name = popname + "_" + part_type
+
+    return dataset_name
+
+
+def is_pop_fluid_file(basename):
+    return (is_particle_file(basename) is False) and "pop" in basename
+
+
+def pop_name(basename):
+    return basename.strip(".h5").split("_")[-2]
+
+
+def make_layout(h5_patch_grp, cell_width):
+    nbrCells = h5_patch_grp.attrs['nbrCells']
+    origin = float(h5_patch_grp.attrs['origin'])
+    upper = int(h5_patch_grp.attrs['upper'])
+    lower = int(h5_patch_grp.attrs['lower'])
+
+    return GridLayout(Box(lower, upper), origin, cell_width)
+
+
+
+def create_from_all_times(time, hier):
+    return time is None and hier is None
+
+
+def create_from_one_time(time, hier):
+    return time is not None and hier is None
+
+
+def load_all_times(time, hier):
+    return time is None and hier is not None
+
+
+def load_one_time(time, hier):
+    return time is not None and hier is not None
+
+
+
+
+def add_to_patchdata(patch_datas, h5_patch_grp, basename, layout):
+    """
+    adds data in the h5_patch_grp in the given PatchData dict
+    """
+
+    if is_particle_file(basename):
+        particles = Particles(icells=h5_patch_grp["iCell"],
+                              deltas=h5_patch_grp["delta"],
+                              v=h5_patch_grp["v"],
+                              weights=h5_patch_grp["weight"],
+                              charges=h5_patch_grp["charge"])
+
+        pdname = particle_dataset_name(basename)
+        if pdname in patch_datas:
+            raise ValueError("error - {} already in patchdata".format(pdname))
+
+        patch_datas[pdname] = ParticleData(layout, particles)
+
+    else:
+        for dataset_name in h5_patch_grp.keys():
+
+            dataset = h5_patch_grp[dataset_name]
+
+            if dataset_name not in field_qties:
+                raise RuntimeError(
+                    "invalid dataset name : {} is not in {}".format(dataset_name, field_qties))
+
+            pdata = FieldData(layout, field_qties[dataset_name], dataset)
+
+            if is_pop_fluid_file(basename):
+                dataset_name = pop_name(basename) + "_" + dataset_name
+
+
+            if dataset_name in patch_datas:
+                raise ValueError("error - {} already in patchdata".format(dataset_name))
+
+            patch_datas[dataset_name] = pdata
+
+
+
+
+
+
+def hierarchy_from(h5_filename, time=None, hier=None):
```

Comment:
what about directory support?

Reply:
see the example in the PR msg

```python
hier = hierarchy_from(diag_path + "EM_E.h5", hier=diag_hier)
```

Reply:
I did, that looks like a file rather than a directory

Reply:
it's the full path

Reply:
yes, is it not useful to also accept a directory and load all files?

Reply:
ah that's what you meant!  Sorry.
yeah maybe.
I'll make an issue ? I don't find it highly important right now. Also often you don't want to read everything because that's just too much.

Reply:
ok as you see fit